yt/frontends/cf_radial/data_structures.py

```diff
@@ -0,0 +1,329 @@
+"""
+CF Radial data structures
+
+
+
+"""
+import contextlib
+import os
+import weakref
+from typing import Optional, Tuple
+
+import numpy as np
+from unyt import unyt_array
+
+from yt.data_objects.index_subobjects.grid_patch import AMRGridPatch
+from yt.data_objects.static_output import Dataset
+from yt.funcs import mylog
+from yt.geometry.grid_geometry_handler import GridIndex
+from yt.utilities.on_demand_imports import _xarray as xr
+
+from .fields import CFRadialFieldInfo
+
+
+class CFRadialGrid(AMRGridPatch):
+    _id_offset = 0
+
+    def __init__(self, id, index, level, dimensions):
+        super().__init__(id, filename=index.index_filename, index=index)
+        self.Parent = None
+        self.Children = []
+        self.Level = level
+        self.ActiveDimensions = dimensions
+
+
+class CFRadialHierarchy(GridIndex):
+    grid = CFRadialGrid
+
+    def __init__(self, ds, dataset_type="cf_radial"):
+        self.dataset_type = dataset_type
+        self.dataset = weakref.proxy(ds)
+        # our index file is the dataset itself:
+        self.index_filename = self.dataset.parameter_filename
+        self.directory = os.path.dirname(self.index_filename)
+        # float type for the simulation edges and must be float64 now
+        self.float_type = np.float64
+        super().__init__(ds, dataset_type)
+
+    def _detect_output_fields(self):
+        # This sets self.field_list, containing all the available on-disk fields and
+        # records the units for each field.
+        self.field_list = []
+        units = {}
+        with self.ds._handle() as xr_ds_handle:
+            for key in xr_ds_handle.variables.keys():
+                if all(x in xr_ds_handle[key].dims for x in ["time", "z", "y", "x"]):
+                    fld = ("cf_radial", key)
+                    self.field_list.append(fld)
+                    units[fld] = xr_ds_handle[key].units
+
+        self.ds.field_units.update(units)
+
+    def _count_grids(self):
+        self.num_grids = 1
+
+    def _parse_index(self):
+        self.grid_left_edge[0][:] = self.ds.domain_left_edge[:]
+        self.grid_right_edge[0][:] = self.ds.domain_right_edge[:]
+        self.grid_dimensions[0][:] = self.ds.domain_dimensions[:]
+        self.grid_particle_count[0][0] = 0
+        self.grid_levels[0][0] = 1
+        self.max_level = 1
+
+    def _populate_grid_objects(self):
+        # only a single grid, no need to loop
+        g = self.grid(0, self, self.grid_levels.flat[0], self.grid_dimensions[0])
+        g._prepare_grid()
+        g._setup_dx()
+        self.grids = np.array([g], dtype="object")
+
+
+class CFRadialDataset(Dataset):
+    _index_class = CFRadialHierarchy
+    _field_info_class = CFRadialFieldInfo
+
+    def __init__(
+        self,
+        filename,
+        dataset_type="cf_radial",
+        storage_filename: Optional[str] = None,
+        storage_overwrite: Optional[bool] = False,
```

Comment:
Just noticed that. Is there a special meaning to `None` here ? I'd assume only booleans are actually valid
```suggestion
        storage_overwrite: bool = False,
```

Reply:
good catch! no, no special meaning